### ISSUE TITLE
Wdio runner does not kill all child processes when killing/interupting wdio

### ISSUE THREAD

**Issue by patthiel** [webdriverio/webdriverio#1549]
## The problem

When running multiple test specs in parallel and attempting to exit wdio (^C) to end the run before through the runner is done running the tests, wdio never exits.. ^C only appears to kill one child process (instance) at a time and not the parent process.

To get around this, i have to continually exit for as many test specs as i  am running to fully exit.

## Environment

* WebdriverIO version: 4.2.5
* Node.js version: 4.2.4
* [wdio testrunner]
* synchronous
* Additional wdio packages used (if applicable):
gulp-webdriver 2.0.2
wdio-jasmine-framework (latest)

## Link to Selenium/WebdriverIO logs

Reproduceable with multiple spec files running in parallel. Expected behavior would be that exiting wdio once ends all processes

@christian-bromann

**Comment by christian-bromann** [comment]
@patthiel can you see if your selenium server received any `end session` requests?

**Comment by patthiel** [comment]
Sure thing. Also, even after killing selenium i get errors from wdio as it's attempting to run the specs still.
[selenium log](https://gist.github.com/patthiel/111bdafaff552c5fcf20f9d417b4b1c4)
[wdio log (maybe less helpful)](https://gist.github.com/patthiel/401e769bd9ea9f73be6e1b6415e4e3b0)

**Comment by christian-bromann** [comment]
Does the same happen if you run without gulp?

**Comment by patthiel** [comment]
It exits better when running without gulp, but I have to exit 3-4 times and this error throws:

```
userName:PROJECT userName$ Error: channel closed
    at process.target.send (internal/child_process.js:510:16)
    at /PROJECT/node_modules/webdriverio/build/lib/runner.js:341:25
    at /PROJECT/node_modules/wdio-jasmine-framework/node_modules/wdio-sync/build/index.js:251:34
Error: channel closed
    at process.target.send (internal/child_process.js:510:16)
    at /PROJECT/node_modules/webdriverio/build/lib/runner.js:329:25
    at /PROJECT/node_modules/wdio-jasmine-framework/node_modules/wdio-sync/build/index.js:251:34

userName:PROJECT userName$ Error: channel closed
    at process.target.send (internal/child_process.js:510:16)
    at /PROJECT/node_modules/webdriverio/build/lib/runner.js:341:25
    at /PROJECT/node_modules/wdio-jasmine-framework/node_modules/wdio-sync/build/index.js:251:34
```

**Comment by dcypherthis** [comment]
A lot of times i have to go for something nuclear like `pkill node`.  ive seen this happen before too.

**Comment by kevinmcdonnell** [comment]
Same, Sometimes I am left with a `nodejs` process running at 100% cpu which I need to kill manually

**Comment by monolithed** [comment]
The problem can be solved by killing Selenium process:

```
➜  ps aux | grep -i selenium 

monolithed      38987   0,0  0,9  6048168  76320   ??  S     3:13     0:20.45 /usr/bin/java -jar /tests/node_modules/selenium-standalone/.selenium/selenium-server/2.53.1-server.jar -Dwebdriver.chrome.driver=/tests/node_modules/selenium-standalone/.selenium/chromedriver/2.23-x64-chromedriver -Dwebdriver.gecko.driver=/tests/node_modules/selenium-standalone/.selenium/geckodriver/0.10.0-x64-geckodriver

➜  kill  38987
```

Btw, `^C` is an unexpected program termination and wdio is not the place for fixing that

**Comment by patthiel** [comment]
@monolithed true.. Although, i think it's a common enough issue for people running tests locally to warrant some invesigation. I'll leave it to @christian-bromann  to decide if this warrants further investigation or if this is out of scope.

**Comment by christian-bromann** [comment]
I think there room for improvements here for sure like killing child process and driver process more aggressively if it won't close with a Selenium command. Not my first priority now.

**Comment by christian-bromann** [comment]
fixed by https://github.com/webdriverio/webdriverio/commit/9ce7f7bb508fb2bb47071c11bbb8b76ff743f0f5

**Comment by monolithed** [comment]
@christian-bromann, could you publish these changes to npm?

**Comment by monolithed** [comment]
```
➜ Error: channel closed
    at process.target.send (internal/child_process.js:540:16)
    at /tests/node_modules/webdriverio/build/lib/runner.js:341:25
    at execHook (/Users/a.abashkin/workspace/mail.ru/e.mail.ru/tests/node_modules/wdio-sync/build/index.js:259:35)
Error: channel closed
    at process.target.send (internal/child_process.js:540:16)
    at /tests/node_modules/webdriverio/build/lib/runner.js:329:25
    at execHook (tests/node_modules/wdio-sync/build/index.js:259:35)
Error: channel closed
    at process.target.send (internal/child_process.js:540:16)
    at /tests/node_modules/webdriverio/build/lib/runner.js:341:25
    at execHook (/tests/node_modules/wdio-sync/build/index.js:259:35)
(node:70944) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 3): Error: channel closed
(node:70223) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Error: channel closed
(node:70005) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Error: channel closed
(node:70280) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Error: channel closed
(node:69964) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Error: channel closed
(node:69925) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Error: channel closed
(node:70043) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Error: channel closed
(node:70317) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Error: channel closed
.....

```

I see this visual noise in node 7, @christian-bromann, what is the problem to publish these changes?

**Comment by goyney** [comment]
I believe this is still an issue in the latest versions. Our terminal windows become flooded with the following errors upon CTRL+C.

```
Error [ERR_IPC_CHANNEL_CLOSED]: channel closed
at process.target.send (internal/child_process.js:590:16)
at /Users/mirigoyen/Git/dtm/auto-tests/node_modules/webdriverio/build/lib/runner.js:484:25
at execHook (/Users/mirigoyen/Git/dtm/auto-tests/node_modules/wdio-sync/build/index.js:166:35)
```

Have to close the terminal window to get it to stop. It makes local code writing and debugging super annoying.

**Comment by monolithed** [comment]
It's because child processes should be replaced with threads/fibers
